File: 프로그래머스/lv1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.py
import math

# 해당 위치를 찾아주는 함수
def getPosition(array, target):
    for i in range(4):
        for j in range(3):
            if array[i][j] == target:
                return i, j

# 두 점 사이의 거리를 구해주는 함수
def getDistance(nowX, nowY, targetX, targetY):
    return abs(nowX - targetX) + abs(nowY - targetY)

def solution(numbers, hand):
    answer = ''
    phone = [[1, 2, 3], [4, 5, 6], [7, 8, 9], ["*", 0, "#"]]
    
    rightHand = [3, 2]
    leftHand = [3, 0]
    
    for number in numbers:
        if number == 1 or number == 4 or number == 7:
            x, y = getPosition(phone, number)
            leftHand[0], leftHand[1] = x, y
            answer += "L"
        elif number == 3 or number == 6 or number == 9:
            x, y = getPosition(phone, number)
            rightHand[0], rightHand[1] = x, y
            answer += "R"
        else:
            x, y = getPosition(phone, number)
            leftHandToTarget = getDistance(leftHand[0], leftHand[1], x, y)
            rightHandToTarget = getDistance(rightHand[0], rightHand[1], x, y)
            if leftHandToTarget < rightHandToTarget:
                leftHand[0], leftHand[1] = x, y
                answer += "L"
            elif leftHandToTarget > rightHandToTarget:
                rightHand[0], rightHand[1] = x, y
                answer += "R"
            else:
                if hand == "left":
                    leftHand[0], leftHand[1] = x, y
                    answer += "L"
                else:
                    rightHand[0], rightHand[1] = x, y
                    answer += "R"
                
    return answer

File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.js
function solution(numbers, hand) {
    
    const findPosition = (target) => {
        const keypads = [[1, 2, 3], [4, 5, 6], [7, 8, 9], ["*", 0, "#"]];
        for (let i = 0; i < 4; i++) {
            for (let j = 0; j < 3; j++) {
                if (keypads[i][j] === target) {
                    return [i, j];
                }
            }
        }
    }
    
    let answer = '';
    const position = {
        L: "*",
        R: "#"
    };
    for (const number of numbers) {
        if (number === 1 || number === 4 || number === 7) {
            answer += "L";
            position.L = number;
        } else if (number === 3 || number === 6 || number === 9) {
            answer += "R";
            position.R = number;
        } else {
            // 거리가 더 가까운 엄지손가락 구하기
            const [currLeftX, currLeftY] = findPosition(position.L);
            const [currRightX, currRightY] = findPosition(position.R);
            const [targetX, targetY] = findPosition(number);
            
            // 거리 구하기
            const leftToTarget = Math.abs(currLeftX - targetX) + Math.abs(currLeftY - targetY);
            const rightToTarget = Math.abs(currRightX - targetX) + Math.abs(currRightY - targetY);
            
            if (leftToTarget < rightToTarget) {
                answer += "L";
                position.L = number;
            } else if (leftToTarget > rightToTarget) {
                answer += "R";
                position.R = number;
            } else {
                hand === "right" ? (position.R = number, answer += "R") : (position.L = number, answer += "L");
            }
        }
    }
    
    return answer;
}